core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java

```diff
@@ -65,4 +83,76 @@ public int totalOperations() {
     public void close() throws IOException {
         onClose.close();
     }
+
+    /**
+     * A wrapper of {@link FixedBitSet} but allows to check if all bits are set in O(1).
+     */
+    private static final class CountedBitSet {
+        private short onBits;
+        private final FixedBitSet bitset;
+
+        CountedBitSet(short numBits) {
+            assert numBits > 0;
+            this.onBits = 0;
+            this.bitset = new FixedBitSet(numBits);
+        }
+
+        boolean getAndSet(int index) {
+            assert index >= 0;
+            boolean wasOn = bitset.getAndSet(index);
+            if (wasOn == false) {
+                onBits++;
+            }
+            return wasOn;
+        }
+
+        boolean hasAllBitsOn() {
+            return onBits == bitset.length();
+        }
+    }
+
+    /**
+     * Sequence numbers from translog are likely to form contiguous ranges,
+     * thus collapsing a completed bitset into a single entry will reduce memory usage.
+     */
+    static final class SeqNumSet {
```

Comment:
Could we call this `SequenceNumberSet` instead?

Reply:
This class is used only here; a concise name in a private context is better. Anyway, I renamed it to `SeqNoSet` as we call `SeqNo*`, not `SeqNum*`.This class is used only here; a concise name in a private context is better. Anyway, I renamed it to `SeqNoSet` as we call `SeqNo*`, not `SeqNum*` in  https://github.com/elastic/elasticsearch/pull/27268/commits/a3994f86d84790051df796868f9fa99ca19a63b6